src/plugins/intel_npu/src/plugin/npuw/compiled_model.cpp

```diff
@@ -406,24 +412,51 @@ ov::npuw::CompiledModel::CompiledModel(const std::shared_ptr<ov::Model>& model,
     reset_io();
 }
 
-void ov::npuw::CompiledModel::fill_weights_bank(const std::size_t idx) {
-    LOG_VERB("Filling weights bank for Subgraph[" << idx << "]...");
-    LOG_BLOCK();
+void ov::npuw::CompiledModel::finalize_weights_bank() {
+    // Evaluate lazy tensors which aren't in the bank beforehand
+    std::vector<std::vector<ov::Tensor>> evaluated_tensors(m_compiled_submodels.size());
+    ov::parallel_for(m_compiled_submodels.size(), [&](std::size_t idx) {
+        evaluated_tensors[idx].resize(m_compiled_submodels[idx].lazy_closure.size());
+        auto& comp_model_desc = m_compiled_submodels[idx];
 
-    NPUW_ASSERT(m_compiled_submodels[idx].replaced_by);
+        if (!comp_model_desc.replaced_by) {
+            return;
+        }
 
-    auto& comp_model_desc = m_compiled_submodels[idx];
+        const auto real_idx = comp_model_desc.replaced_by.value();
+        auto& func_desc = m_compiled_submodels[real_idx];
 
-    for (std::size_t cidx = 0u; cidx < comp_model_desc.closure.size(); cidx++) {
-        comp_model_desc.closure[cidx] = m_weights_bank->update(comp_model_desc.closure[cidx]);
-        if (m_cfg.get<::intel_npu::NPUW_FOLD>()) {
-            comp_model_desc.update_required[cidx] = true;
-        } else {
-            comp_model_desc.update_required[cidx] = false;
+        for (std::size_t tidx = 0; tidx < comp_model_desc.lazy_closure.size(); ++tidx) {
+            const auto& lt = m_compiled_submodels[idx].lazy_closure[tidx];
+            if (m_weights_bank->has(lt, *func_desc.device_it)) {
+                continue;
+            }
+            evaluated_tensors[idx][tidx] = lt.eval();
         }
-    }
+    });
 
-    LOG_VERB("DONE");
+    for (size_t idx = 0; idx < m_compiled_submodels.size(); ++idx) {
+        auto& comp_model_desc = m_compiled_submodels[idx];
+
+        if (!comp_model_desc.replaced_by) {
+            continue;
+        }
+
+        const auto real_idx = comp_model_desc.replaced_by.value();
+        auto& func_desc = m_compiled_submodels[real_idx];
+
+        // Due to concat some tensor should be skipped in closure
+        m_compiled_submodels[idx].closure.resize(0);
+        m_compiled_submodels[idx].is_remote.resize(0);
+
+        for (std::size_t tidx = 0; tidx < comp_model_desc.lazy_closure.size(); ++tidx) {
+            const auto& lt = m_compiled_submodels[idx].lazy_closure[tidx];
+            const auto& evaled = evaluated_tensors[idx][tidx];
+            m_compiled_submodels[idx].closure.push_back(m_weights_bank->get(lt, *func_desc.device_it, evaled));
+            // FIXME: should is_remote be set unconditionally?
+            m_compiled_submodels[idx].is_remote.push_back(true);
```

Comment:
@dmatveev it works even on CPU. Is it still required for any other case? (perhaps w/o folding)

Reply:
Done, thanks